File: app/src/main/java/com/trae/agent/llm/LlmClient.java
package com.trae.agent.llm;

import com.trae.agent.model.Message;
import com.trae.agent.model.ToolCall;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * OpenAI-compatible LLM client.
 * Supports chat completions with function/tool calling.
 */
public class LlmClient {

    private final LlmConfig config;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public LlmClient(LlmConfig config) {
        this.config = config;
    }

    /**
     * Send a chat completion request synchronously.
     *
     * @param messages  conversation history
     * @param toolsJson JSON array of tool definitions (OpenAI tool format), may be null
     * @return parsed LLM response
     */
    public LlmResponse chat(List<Message> messages, JSONArray toolsJson) throws IOException {
        JSONObject body = new JSONObject();
        body.put("model", config.getModel());
        body.put("temperature", config.getTemperature());
        body.put("max_tokens", config.getMaxTokens());
        body.put("messages", messagesToJson(messages));
        if (toolsJson != null && toolsJson.length() > 0) {
            body.put("tools", toolsJson);
        }

        String responseJson = postJson(config.getBaseUrl() + "/chat/completions", body.toString());

        return parseResponse(responseJson);
    }

    /**
     * Async variant.
     */
    public Future<LlmResponse> chatAsync(List<Message> messages, JSONArray toolsJson) {
        return executor.submit(() -> chat(messages, toolsJson));
    }

    // ---- HTTP ----

    private String postJson(String urlStr, String jsonBody) throws IOException {
        HttpURLConnection conn = (HttpURLConnection) new URL(urlStr).openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setRequestProperty("Authorization", "Bearer " + config.getApiKey());
        conn.setDoOutput(true);
        conn.setConnectTimeout(30_000);
        conn.setReadTimeout(120_000);

        try (OutputStream os = conn.getOutputStream()) {
            os.write(jsonBody.getBytes(StandardCharsets.UTF_8));
        }

        int status = conn.getResponseCode();
        BufferedReader reader;
        if (status >= 200 && status < 300) {
            reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        } else {
            reader = new BufferedReader(new InputStreamReader(conn.getErrorStream(), StandardCharsets.UTF_8));
        }

        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line);
        }
        reader.close();

        if (status < 200 || status >= 300) {
            throw new IOException("LLM API error " + status + ": " + sb);
        }
        return sb.toString();
    }

    // ---- JSON helpers ----

    private JSONArray messagesToJson(List<Message> messages) {
        JSONArray arr = new JSONArray();
        for (Message msg : messages) {
            JSONObject obj = new JSONObject();
            obj.put("role", msg.getRole());
            obj.put("content", msg.getContent() != null ? msg.getContent() : "");

            // tool_calls from assistant
            if (msg.getToolCalls() != null && !msg.getToolCalls().isEmpty()) {
                JSONArray tcs = new JSONArray();
                for (ToolCall tc : msg.getToolCalls()) {
                    JSONObject tcObj = new JSONObject();
                    tcObj.put("id", tc.getId());
                    tcObj.put("type", tc.getType());
                    JSONObject func = new JSONObject();
                    func.put("name", tc.getName());
                    func.put("arguments", tc.getArguments() != null
                            ? new JSONObject(tc.getArguments()).toString()
                            : "{}");
                    tcObj.put("function", func);
                    tcs.put(tcObj);
                }
                obj.put("tool_calls", tcs);
            }

            // tool call result
            if ("tool".equals(msg.getRole())) {
                obj.put("tool_call_id", msg.getToolCallId());
            }

            arr.put(obj);
        }
        return arr;
    }

    private LlmResponse parseResponse(String json) {
        JSONObject root = new JSONObject(json);
        JSONObject choice = root.getJSONArray("choices").getJSONObject(0);
        JSONObject messageJson = choice.getJSONObject("message");

        String role = messageJson.optString("role", "assistant");
        String content = messageJson.optString("content", null);

        LlmResponse response = new LlmResponse(role, content);

        if (messageJson.has("tool_calls")) {
            JSONArray tcs = messageJson.getJSONArray("tool_calls");
            List<ToolCall> toolCalls = new ArrayList<>();
            for (int i = 0; i < tcs.length(); i++) {
                JSONObject tc = tcs.getJSONObject(i);
                JSONObject func = tc.getJSONObject("function");
                String argsStr = func.optString("arguments", "{}");

                Map<String, Object> args = new HashMap<>();
                try {
                    JSONObject argsJson = new JSONObject(argsStr);
                    for (String key : argsJson.keySet()) {
                        args.put(key, argsJson.get(key));
                    }
                } catch (Exception ignored) {}

                ToolCall toolCall = new ToolCall(
                        tc.getString("id"),
                        func.getString("name"),
                        args
                );
                toolCalls.add(toolCall);
            }
            response.setToolCalls(toolCalls);
        }

        // token usage
        if (root.has("usage")) {
            JSONObject usage = root.getJSONObject("usage");
            response.setPromptTokens(usage.optInt("prompt_tokens", 0));
            response.setCompletionTokens(usage.optInt("completion_tokens", 0));
        }

        response.setFinishReason(choice.optString("finish_reason", "stop"));
        return response;
    }

    // ---- Response DTO ----

    public static class LlmResponse {
        private final String role;
        private final String content;
        private List<ToolCall> toolCalls;
        private int promptTokens;
        private int completionTokens;
        private String finishReason;

        public LlmResponse(String role, String content) {
            this.role = role;
            this.content = content;
        }

        public String getRole() { return role; }
        public String getContent() { return content; }
        public List<ToolCall> getToolCalls() { return toolCalls; }
        public void setToolCalls(List<ToolCall> toolCalls) { this.toolCalls = toolCalls; }
        public int getPromptTokens() { return promptTokens; }
        public void setPromptTokens(int promptTokens) { this.promptTokens = promptTokens; }
        public int getCompletionTokens() { return completionTokens; }
        public void setCompletionTokens(int completionTokens) { this.completionTokens = completionTokens; }
        public String getFinishReason() { return finishReason; }
        public void setFinishReason(String finishReason) { this.finishReason = finishReason; }
        public boolean hasToolCalls() { return toolCalls != null && !toolCalls.isEmpty(); }
    }
}

File: codingAgent/src/llm/llmClient.js
"use strict";

const axios = require("axios");
const config = require("../config");

/**
 * 通用的 LLM 客户端，默认兼容 OpenAI API 格式。
 * 支持多轮对话 chat 与 embedding 两种核心能力。
 */
class LlmClient {
    /**
     * 构造 LLM 客户端，使用 config.llm 中的配置。
     */
    constructor() {
        this.provider = config.llm.provider;
        this.apiKey = config.llm.apiKey;
        this.baseUrl = config.llm.baseUrl.replace(/\/$/, "");
        this.model = config.llm.model;
        this.embeddingModel = config.llm.embeddingModel;
        this.maxRetries = config.llm.maxRetries;
        this.timeoutMs = config.llm.timeoutMs;
    }

    /**
     * 发起聊天补全请求。
     * @param {Array<{role:string, content:string, name?:string}>} messages 消息列表
     * @param {Object} [options={}] 额外参数，如 tools、tool_choice、temperature
     * @returns {Promise<Object>} LLM 返回的完整 choice 对象
     */
    async chat(messages, options) {
        if (!options) {
            options = {};
        }

        const url = `${this.baseUrl}/chat/completions`;
        const payload = {
            model: this.model,
            messages: messages
        };

        if (options.tools) {
            payload.tools = options.tools;
        }
        if (options.tool_choice) {
            payload.tool_choice = options.tool_choice;
        }
        if (typeof options.temperature === "number") {
            payload.temperature = options.temperature;
        }

        let lastError = null;
        for (let attempt = 0; attempt < this.maxRetries; attempt += 1) {
            try {
                const response = await axios.post(url, payload, {
                    headers: this._buildHeaders(),
                    timeout: this.timeoutMs
                });
                const choices = response.data && response.data.choices;
                if (!choices || choices.length === 0) {
                    throw new Error("LLM 返回 choices 为空");
                }
                return choices[0];
            } catch (error) {
                lastError = error;
                if (attempt < this.maxRetries - 1) {
                    const delayMs = 1000 * Math.pow(2, attempt);
                    await this._sleep(delayMs);
                }
            }
        }

        throw new Error(`LLM 请求失败（已重试 ${this.maxRetries} 次）: ${this._formatError(lastError)}`);
    }

    /**
     * 获取文本的 embedding 向量，用于 RAG 检索。
     * @param {string|string[]} input 单条或多条文本
     * @returns {Promise<number[][]>} 向量列表
     */
    async embed(input) {
        if (typeof input === "string") {
            input = [input];
        }

        const url = `${this.baseUrl}/embeddings`;
        const payload = {
            model: this.embeddingModel,
            input: input
        };

        let lastError = null;
        for (let attempt = 0; attempt < this.maxRetries; attempt += 1) {
            try {
                const response = await axios.post(url, payload, {
                    headers: this._buildHeaders(),
                    timeout: this.timeoutMs
                });
                const data = response.data && response.data.data;
                if (!data || data.length === 0) {
                    throw new Error("Embedding 返回数据为空");
                }
                return data.map(function (item) {
                    return item.embedding;
                });
            } catch (error) {
                lastError = error;
                if (attempt < this.maxRetries - 1) {
                    const delayMs = 1000 * Math.pow(2, attempt);
                    await this._sleep(delayMs);
                }
            }
        }

        throw new Error(`Embedding 请求失败（已重试 ${this.maxRetries} 次）: ${this._formatError(lastError)}`);
    }

    /**
     * 构造 HTTP 请求头，注入认证信息。
     * @returns {Object} headers
     */
    _buildHeaders() {
        const headers = {
            "Content-Type": "application/json"
        };
        if (this.apiKey) {
            headers.Authorization = `Bearer ${this.apiKey}`;
        }
        return headers;
    }

    /**
     * 将错误对象格式化为可读的字符串。
     * @param {Error} error
     * @returns {string}
     */
    _formatError(error) {
        if (!error) {
            return "未知错误";
        }
        if (error.response && error.response.data) {
            return `${error.message}; 响应: ${JSON.stringify(error.response.data)}`;
        }
        return error.message;
    }

    /**
     * 等待指定毫秒。
     * @param {number} ms
     * @returns {Promise<void>}
     */
    _sleep(ms) {
        return new Promise(function (resolve) {
            setTimeout(resolve, ms);
        });
    }
}

module.exports = LlmClient;
